app/uk/gov/hmrc/tariffclassificationfrontend/controllers/CaseController.scala

```diff
@@ -119,23 +138,37 @@ class CaseController @Inject()(authenticatedAction: AuthenticatedAction,
     activityForm.bindFromRequest.fold(onError, onSuccess)
   }
 
-  private def showActivity(c: Case, f: Form[ActivityFormData])
-                          (implicit request: AuthenticatedRequest[AnyContent]): Future[HtmlFormat.Appendable] = {
-    for {
-      events <- eventsService.getEvents(c.reference, NoPagination())
-      queues <- queuesService.getAll
-    } yield views.html.partials.activity_details(c, events, f, queues)
+  private def getCaseAndRedirect(reference: String, page: CaseDetailPage, toHtml: Case => Future[Call])
+                                (implicit request: Request[_]): Future[Result] = {
+    casesService.getOne(reference).flatMap {
+      case Some(c: Case) => toHtml(c).map(_ => Redirect(routes.CaseController.activityDetails(reference)))
+      case _ => successful(Ok(views.html.case_not_found(reference)))
+    }
   }
 
-  def keywordsDetails(reference: String): Action[AnyContent] = authenticatedAction.async { implicit request =>
+  def keywordsDetails(reference: String): Action[AnyContent] = verify.caseExistsAndCheckPermissions(reference).async { implicit request =>
     getCaseAndRenderView(
       reference,
       KEYWORDS,
-      showKeywords(_, keywordForm)
+      showKeywords(_, keywordForm),
+      request.c
     )
   }
 
-  def addKeyword(reference: String): Action[AnyContent] = authenticatedAction.async { implicit request =>
+  private def showKeywords(c: Case, f: Form[String])
+                          (implicit request: AuthenticatedRequest[AnyContent]): Future[HtmlFormat.Appendable] = {
+    keywordsService.autoCompleteKeywords.map { keywords: Seq[String] =>
+      views.html.partials.keywords_details(c, keywords, f)
+    }
+  }
+
+  private def getCaseAndRenderView(reference: String, page: CaseDetailPage, toHtml: Case => Future[Html], c: Case)
+                                  (implicit request: Request[_]): Future[Result] = {
+
+    toHtml(c).map(html => Ok(views.html.case_details(c, page, html)))
+  }
+
+  def addKeyword(reference: String): Action[AnyContent] = verify.caseExistsAndFilterByAuthorisation(reference).async { implicit request =>
```

Comment:
`caseExistsAndFilterByAuthorisation`
How do we know what that *actually* does without drilling into the action?
 Do you not think it would be easier to understand as something like:
 `verify.caseExists(ref) andThen verify.permittedToModify(Permission.KEYWORD)` or similar